### PR TITLE
io: Error on over-long read()/write().

### LINKED ISSUE #18845: `json.dump` behaves incorrectly on invalid stream write bytes count.
### Port, board and/or hardware

Unix

### MicroPython version

MicroPython v1.28.0-preview.174.ga8b71559cf on 2026-02-20; linux [GCC 15.2.1] version

### Reproduction

1. Run the following minimised code snippet:

```py
import io, json


class S(io.IOBase):
    def write(self, buf):
        return 2


json.dump({}, S())
```

### Expected behaviour

The code should either reject stream writes with an unexpected number of bytes written or just ignore them.

### Observed behaviour

The interpreter hangs due to repeated calls to `io.IOBase.write`, and after a while it will crash on certain targets.

### Additional Information

CPython seems to ignore the return value of the stream callback and only checks for exceptions being raised.  The test case code doesn't hang or crash on all recent CPython versions I've tried.

I understand this is an edge case that may not be worth fixing, but I didn't find anything similar in the issue tracker or in the known issues list present in the documentation.

### Code of Conduct

Yes, I agree

### PR DESCRIPTION
**Pull request by jepler**
### Summary

It is a mistake in Python code if the result of readinto()/write() is more bytes than requested. Before this PR, incorrect values would lead to misbehavior as reported in #18845 if this invariant is not respected.

CPython appears to largely ignore the values returned from readinto()/write(), at least from `print()` & `json.load()`. Consequently, an expected output file is needed for the new test.

Closes: #18845

### Testing

I implemented a new test and ran it on the unix coverage build.

### Trade-offs and Alternatives

It seemed possible that this could go down in mp_stream_rw instead, but any core implementations of stream can be assumed well-behaved. It seemed better to keep the check closest to where the wrong value could be returned from user-controlled code.

### Generative AI

I did not use generative AI tools when creating this PR.

### CODE REVIEW

File: py/modio.c [hunk 1]
```diff
@@ -62,6 +62,9 @@ static mp_uint_t iobase_read_write(mp_obj_t obj, void *buf, mp_uint_t size, int
         return MP_STREAM_ERROR;
     }
     mp_int_t ret = mp_obj_get_int(ret_obj);
+    if (ret > (mp_int_t)size) {
```
Comment (dpgeorge) [inline py/modio.c]: Suggest putting this check down within the `if (ret >= 0)` block.  That will be a tiny bit more efficient for cases which are an error, and allow `ret` to be cast to `(mp_uint_t)` which prevents a (very unlikely) failure in case `size` becomes negative when casting to `mp_int_t`.

File: tests/micropython/io_badlength.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+try:
```
Comment (dpgeorge) [inline tests/micropython/io_badlength.py:4]: Please add a comment along the lines of: Test when a user IOBase class has write/readinto that returns more data than requested.

File: py/modio.c (hunk 1 above)
Comment (jepler) [reply py/modio.c]: Thanks, updated. It looks like this alternative takes +1 additional bytes on ports/unix VARIANT=standard (x86_64) but +0 on ports/qemu BOARD=MICROBIT.

**Review by dpgeorge — APPROVED** [review]

File: py/modio.c (hunk 1 above)
Comment (dpgeorge) [reply py/modio.c]: Thanks.

According to the code-size-diff CI (which for some reason didn't update the comment above...), rp2 and samd both change by -8 bytes with this change.  So that's good at least.

**Comment by dpgeorge** [comment]
Rebased and merged in 5b127f022e5e22717cd013820c9fd76ab90e1e72
